numpy/tests/test_public_api.py

```diff
@@ -69,6 +70,23 @@ def test_numpy_namespace():
     assert bad_results == whitelist
 
 
+@pytest.mark.parametrize('name', ['testing', 'Tester'])
+def test_import_testing(name):
+    # Make sure we can actually access `np.testing`
+    # While not exported as part of the public API, it was accessible.
+    # With the use of __getattr__ and __dir__, this isn't always true
+    # It can happen that an infinite recursion may happen.
+    # This is the only way I found to avoid avoid it
+    sub = Popen((sys.executable, '-c', "import numpy; numpy." + name),
+                stdout=PIPE, stderr=PIPE)
+
+    # on failure, this will produce alot of output, and potentially
+    # deadlock the call without a call to `communicate`
+    sub.communicate()
+    sub.wait()
```

Comment:
Why not just `subprocess.check_call(...)`, which handles all this for you?

Reply:
On failure, you would get some output that seems quite large, all looking like 

<details>

```
  File "<frozen importlib._bootstrap>", line 1032, in _handle_fromlist
  File "/home/mark/git/numpy/numpy/__init__.py", line 198, in __getattr__
    from . import testing
  File "<frozen importlib._bootstrap>", line 1032, in _handle_fromlist
  File "/home/mark/git/numpy/numpy/__init__.py", line 198, in __getattr__
    from . import testing
  File "<frozen importlib._bootstrap>", line 1032, in _handle_fromlist
  File "/home/mark/git/numpy/numpy/__init__.py", line 198, in __getattr__
    from . import testing
  File "<frozen importlib._bootstrap>", line 1032, in _handle_fromlist
  File "/home/mark/git/numpy/numpy/__init__.py", line 198, in __getattr__
    from . import testing
  File "<frozen importlib._bootstrap>", line 1032, in _handle_fromlist
RecursionError: maximum recursion depth exceeded
```
</details>

I think you end up hitting 

> Warning The preexec_fn parameter is not safe to use in the presence of threads in your application. The child process could deadlock before exec is called. If you must use it, keep it trivial! Minimize the number of libraries you call into. 

I think the python 2.7 docs explain it better than the 3.7 docs, and suggests doing what I implemented above
https://docs.python.org/2/library/subprocess.html#subprocess.check_output

Reply:
> output that seems quite large, 

This output goes on the testing console, quite annoying. Therefore I opted to `PIPE` the output. I guess I could also send it to a tempfile???

Reply:
If your goal is to discard it, `check_call(stdout=
subprocess.DEVNULL, stderr=
subprocess.DEVNULL)` should do the job .

I'd maybe argue that the output contains useful debugging information, and pytest will discard it unless it fails anyway - so why not keep it around and use `check_call` with no extra arguments?

Reply:
Pytest is not catching the error output and it is going straight to the consule when that one specific test is run

Reply:
hmm, ok, i found a way.

Reply:
Windows seems to have issues with piping stderr to stdout. Since this will only output stuff during a failure, I think it is acceptable to leave it undefined without complicating the code. Hopefully the test is more clear now.